lib/experimental/auto-inserting-blocks.php

```diff
@@ -0,0 +1,246 @@
+ */
+function gutenberg_register_auto_inserted_blocks( $settings, $metadata ) {
+	if ( ! isset( $metadata['__experimentalAutoInsert'] ) ) {
+		return $settings;
+	}
+	$auto_insert = $metadata['__experimentalAutoInsert'];
+
+	/**
+	 * Map the camelCased position string from block.json to the snake_cased block type position
+	 * used in the auto-inserting block registration function.
+	 *
+	 * @var array
+	 */
+	$property_mappings = array(
+		'before'     => 'before',
+		'after'      => 'after',
+		'firstChild' => 'first_child',
+		'lastChild'  => 'last_child',
+	);
+
+	$inserted_block_name = $metadata['name'];
+	foreach ( $auto_insert as $anchor_block_name => $position ) {
+		// Avoid infinite recursion (auto-inserting next to or into self).
+		if ( $inserted_block_name === $anchor_block_name ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Cannot auto-insert block next to itself.', 'gutenberg' ),
+				'6.4.0'
+			);
+			continue;
+		}
+
+		if ( ! isset( $property_mappings[ $position ] ) ) {
+			continue;
+		}
+
+		$mapped_position = $property_mappings[ $position ];
+
+		gutenberg_register_auto_inserted_block( $inserted_block_name, $mapped_position, $anchor_block_name );
+
+		$settings['auto_insert'][ $anchor_block_name ] = $mapped_position;
+	}
+
+	return $settings;
+}
+add_filter( 'block_type_metadata_settings', 'gutenberg_register_auto_inserted_blocks', 10, 2 );
+
+/**
+ * Register block for auto-insertion into the frontend and REST API.
+ *
+ * Register a block for auto-insertion into the frontend and into the markup
+ * returned by the templates and patterns REST API endpoints.
+ *
+ * This is currently done by filtering parsed blocks as obtained from a block template
+ * template part, or pattern and injecting the auto-inserted block where applicable.
+ *
+ * @todo In the long run, we'd likely want some sort of registry for auto-inserted blocks.
+ *
+ * @param string $inserted_block  The name of the block to insert.
+ * @param string $position        The desired position of the auto-inserted block, relative to its anchor block.
+ *                                Can be 'before', 'after', 'first_child', or 'last_child'.
+ * @param string $anchor_block    The name of the block to insert the auto-inserted block next to.
+ * @return void
+ */
+function gutenberg_register_auto_inserted_block( $inserted_block, $position, $anchor_block ) {
+		$inserted_block = array(
+			'blockName'    => $inserted_block,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+
+		$inserter = gutenberg_auto_insert_block( $inserted_block, $position, $anchor_block );
+		add_filter( 'gutenberg_serialize_block', $inserter, 10, 1 );
+}
+
+/**
+ * Parse and reserialize block templates to allow running filters.
+ *
+ * By parsing a block template's content and then reserializing it
+ * via `gutenberg_serialize_blocks()`, we are able to run filters
+ * on the parsed blocks.
```

Comment:
we spoke about this, but given that what we're talking about here is possible without parsing and serializing I could see it being useful to explain why we're going out of our way to avoid doing that. we could say we're taking advantage of the depth-first block traversal already provided by the serialization process, for instance.

also it's unclear why this makes it possible to run filters, why it's necessary to do so, and why we want to. a counter-example showing why the simpler process doesn't work might prevent someone from accidentally breaking it.

Reply:
Thank you, very good points. First pass: 9c58e8710a

Reply:
> also it's unclear why this makes it possible to run filters, why it's necessary to do so, and why we want to. 

As for this, I tried to cover that in a comment on top of the `gutenberg_serialize_block[s]` functions: https://github.com/WordPress/gutenberg/blob/4ea0b5f4b59b234e114da8f290a10e36863c51fb/lib/experimental/auto-inserting-blocks.php#L185-L189

I tried to avoid referencing the specific use case here (and instead just said "entry point for modifications to the parsed blocks"), since in my mind, the addition of the hook to `gutenberg_serialize_block[s]` was more general-purpose, whereas the specific reason should be mentioned only in the filters we are actually adding to that hook.

I'm open to adding some more specific examples to either though 🤔

Reply:
Minor tweak to `gutenberg_parse_and_serialize_block_templates`' and `gutenberg_parse_and_serialize_blocks`'s PHPDoc in 2446e3d923.